add a progress notification when performing any git clone

olexii4 opened the pull request:
Add a progress notification when performing any git clone.

![screenshot from 2019-03-05 18-25-02](https://user-images.githubusercontent.com/6310786/53820255-0e4a0900-3f74-11e9-8444-5cb7464ec090.png)

Related issue: https://github.com/eclipse/che/issues/12564

Signed-off-by: Oleksii Orel <oorel@redhat.com>
<!-- Please provide a clear and meaningful description to the CHANGELOG.md file if this PR contributes some significant changes -->

akosyakov commented:
cc @svenefftinge @kittaakos

vinokurig commented:
It is strange for me that the notification can be closed by itself, even if clone is not finished yet. The button that restores the notification looks not obvious, I mean it is not clear what it does.

olexii4 commented:
@vinokurig I add it like a proposal. Because it will be nice for me to hide an annoying notification after 10s.

olexii4 commented:
![screenshot from 2019-03-06 15-25-45](https://user-images.githubusercontent.com/6310786/53904388-9dc3ea80-404e-11e9-8477-391d40522689.png)
![screenshot from 2019-03-06 15-31-08](https://user-images.githubusercontent.com/6310786/53904397-a4526200-404e-11e9-9443-0cb74e85870d.png)

olexii4 commented:
WDYT?

svenefftinge commented:
Yes, why not. Ideally we would do the clone in the terminal and rely on the natural output of git.
@AlexTugarev could you have a look at the changes regarding the notifications?

AlexTugarev commented:
I really like the idea to show the progress git clone, actually of all kind of long running operations. However I think the UX didn't work in my test: 

## Starting a git clone operation
It shows `Cloning the git repo` progress notification 👍 
<img width="1234" alt="screen shot 2019-03-07 at 08 20 18" src="https://user-images.githubusercontent.com/914497/53939457-5face200-40b3-11e9-85fe-e2b89b7ef4e5.png">

The progress notification disappears after a short while, but I see spinner icon in the status bar for just two seconds. 

## Starting a second clone operation
... of a git repo containing a lot of objects.
![2019-03-07 08 16 38](https://user-images.githubusercontent.com/914497/53939631-fa0d2580-40b3-11e9-8752-cbd58a9b2008.gif)

And now I see purpose of the status bar item. I says `Tasks in Progress 1/1` and it reveals the progress notification which was hidden previously. 

It's not clear why the progress is hidden in the first place. Especially the progress notifications are meant to report progress continuously:
https://github.com/theia-ide/theia/blob/56222486fd076d3d71887e5e629152015faf88e5/packages/core/src/common/message-service-protocol.ts#L61-L71


And we can easily add the progress reporting from dugite-extra in the Git interface, see [ICloneProgress](https://github.com/theia-ide/dugite-extra/blob/2e03d551cc8d939a9f1d450ac458cfe5f5d16a20/src/command/clone.ts#L41)
 which is to be add here:
https://github.com/theia-ide/theia/blob/56222486fd076d3d71887e5e629152015faf88e5/packages/git/src/node/dugite-git.ts#L339

@olexii4 could you explain what the benefit of the status bar item is? Also, have you seen https://github.com/theia-ide/theia/pull/2461? The generic progress reporting mechanism seems very similar.

olexii4 commented:
@AlexTugarev  
I have fixed the error and want to explain the benefit of the status bar item:
For this moment we have a possibility to show progress message with possible actions to the user.   But, in general case, the progress message is shown under the command palette.  And we should close all notifications for working with the command palette.
But, in the case when we close the existing progress message - lost the opportunity to know when the progress will be canceled.

With my improvement, we close any progress message in time from the preferences(notification.timeout). Then can reopen with the status bar widget.

And I added minimal changes to improve existing progress message, to become more convenient to users because we are using it with our API.  And this improvement was added in the area of the issue "add progress when performing any git clone" like a separate commit to improving their behavior.

Progress extension looks very nice.  But still not merged from the 2018 summer season.  And the big improvement cannot be done in the issue of adding existing progress notification for git clone.

I can create The separate issue for progress message improvement when the progress extension will be merged. Like a part for withProgress improvement(theia.d.ts).

AlexTugarev commented on packages/messages/src/browser/notifications-message-client.ts:
```diff
@@ -48,21 +52,59 @@ export class NotificationsMessageClient extends MessageClient {
     showProgress(progressId: string, message: ProgressMessage, cancellationToken: CancellationToken, update?: ProgressUpdate): Promise<string | undefined> {
-        const messageArguments = { ...message, type: MessageType.Progress, options: { ...(message.options || {}), timeout: 0 } };
-        if (this.visibleProgressNotifications.has(progressId)) {
+        const messageArguments = {
```
why should the default value for `timeout` be changed now?

olexii4 replied:
Because I understand the problem - In the case when you call all notifications from the status bar widget, it is not enough time to read.
You can close it manually with the 'close' button.
We are using default timeout the first time only

olexii4 replied:
https://github.com/theia-ide/theia/blob/4382a27f709c60cacfe1113534fda0f2e7d0d80e/packages/messages/src/browser/notifications-message-client.ts#L160

olexii4 replied:
if we remove ```, timeout: 0```  it will get it from  preferences['notification.timeout']

AlexTugarev replied:
got it. the question is, why should that be the case in general? what is wrong about it in general? besides the git clone operation, the main use case when introducing this feature was to allow progress reporting. what happens with that now?

if the source for progress reports stops updating, will this dismiss the progress notification? 

if you think this should be a short timeout value for git clone in this PR, why not just setting it explicitly?

olexii4 replied:
I cannot understand you.  Previously it was shown without a close timeout at all ``` timeout: 0 ```. 

If the source for progress reports stops updating(as I understand - cancel this progress notification
https://github.com/theia-ide/theia/pull/4483/files#diff-cde0e8b29563cef99910b9e41f4daab7R89
), will this dismiss the progress notification.

AlexTugarev commented:
> But, in general case, the progress message is shown under the command palette. And we should close all notifications for working with the command palette.

The problem is, that the UI for notification is worn out and [needs a refresh](https://github.com/theia-ide/theia/issues/1360). 

But here, a fix of the z-order of the notification would solve the issue, right? If you still can use the command palette while the progress notification is visible, you don't need to dismiss it automatically.

> With my improvement, we close any progress message in time from the preferences(notification.timeout). 

What is the reason to change this default?

> Progress extension looks very nice. But still not merged from the 2018 summer season. And the big improvement cannot be done in the issue of adding existing progress notification for git clone.

Well, isn't it worth spending time on finishing stuff? The progress UI just had [no (well) working](https://github.com/theia-ide/theia/pull/2461#issuecomment-435843176) application. The git clone would be a good candidate, especially with the hints about the progress reporting from above. I'd rather see progress extension redux using the git clone case as an application of it, instead of adding yet another shortcut. 

I'm releasing the break as the memory leak is gone. But please get more people convinced of the UX.

AlexTugarev commented on packages/git/src/browser/git-quick-open-service.ts:
```diff
@@ -77,11 +77,16 @@ export class GitQuickOpenService {
                     dynamicItems.push(new SingleStringInputOpenItem(
                         `Clone the Git repository: ${lookFor}. ${suffix}`,
                         async () => {
+                            const progress = await gitQuickOpenService.messageService.showProgress({
```
cf. links in:
https://github.com/eclipse/che/issues/12564#issuecomment-459344603
https://github.com/eclipse/che/issues/12564#issuecomment-459344937

it was discussed couple of times before. and it would really make sense so tackle this in a more generic way.

olexii4 replied:
Ok. There's more general issue https://github.com/theia-ide/theia/issues/759.

What about approve improvements for inProgress notifications only?
We are using them in our API and it will be nice to improve the existing behavior.  In this case, I will remove the first commit and change the PR description

olexii4 commented:
@AlexTugarev
Could we discuss the second commit for inProgress notification improvements only?  I am sure it will be a good improvement
